types/src/index.ts

```diff
@@ -0,0 +1,102 @@
+#!/usr/bin/env node
+import assert from "assert";
+import { mkdir, readFile, writeFile } from "fs/promises";
+import path from "path";
+import { arrayBuffer } from "stream/consumers";
+import util from "util";
+import { StructureGroups } from "@workerd/jsg/rtti.capnp.js";
+import { Message } from "capnp-ts";
+import prettier from "prettier";
+import ts from "typescript";
+import { generateDefinitions } from "./generator";
+import { printNodeList, printer } from "./print";
+import { createMemoryProgram } from "./program";
+import {
+  createGlobalScopeTransformer,
+  createIteratorTransformer,
+} from "./transforms";
+
+const definitionsHeader = `/* eslint-disable */
+// noinspection JSUnusedGlobalSymbols
+`;
+
+function printDefinitions(root: StructureGroups): string {
+  // Generate TypeScript nodes from capnp request
+  const nodes = generateDefinitions(root);
+
+  // Build TypeScript program from nodes
+  const source = printNodeList(nodes);
+  // TODO(soon): when we switch to outputting a separate file per group, we'll
+  //  need to modify this function to accept multiple source files
+  //  (will probably need `program.getSourceFiles()`)
+  const [program, sourcePath] = createMemoryProgram(source);
+  const checker = program.getTypeChecker();
+  const sourceFile = program.getSourceFile(sourcePath);
+  assert(sourceFile !== undefined);
+
+  // Run post-processing transforms on program
+  const result = ts.transform(sourceFile, [
+    // TODO(soon): when overrides are implemented, apply renames here
+    createIteratorTransformer(checker),
+    createGlobalScopeTransformer(checker),
+    // TODO(polish): maybe flatten union types?
+  ]);
+  // TODO(polish): maybe log diagnostics with `ts.getPreEmitDiagnostics(program, sourceFile)`?
+  //  (see https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API#a-minimal-compiler)
+  assert.strictEqual(result.transformed.length, 1);
+
+  // Print program to string
+  return definitionsHeader + printer.printFile(result.transformed[0]);
+}
+
+// Generates TypeScript types from a binary Cap’n Proto file containing encoded
+// JSG RTTI. See src/workerd/tools/api-encoder.c++ for a script that generates
+// input expected by this tool.
+//
+// To generate types using default options, run `bazel build //types:types`.
+//
+// Usage: types [options] [input]
+//
+// Options:
+//  -o, --output <file>
+//    File path to write TypeScript to, defaults to stdout if omitted
+//  -f, --format
+//    Formats generated types with Prettier
+//
+// Input:
+//    Binary Cap’n Proto file path, defaults to reading from stdin if omitted
+export async function main(args?: string[]) {
+  const { values: options, positionals } = util.parseArgs({
```

Comment:
I would just caution here that `util.parseArgs()` is still considered experimental in Node.js. It's API can see breaking changes outside of normal semver rules. It's not expected to change, but it's worth keeping in mind.